File: 8. String to Integer (atoi)/8. String to Integer (atoi).py
class Solution:
    def myAtoi(self, str: 'str') -> 'int':
        if not str:
            return 0
        
        str = str.strip() # remve whitespaces
        if len(str) < 1:
            return 0
        
        sign = 1
        idx = 0
        if str[0] == "-":
            sign = -1
            idx = 1
        elif str[0] == "+":
            sign = sign
            idx = 1
        
        num = 0
        for n in str[idx:]:
            if n not in "1234567890":
                break
            num = 10*num + ord(n) - ord("0")

        num = sign*int(num)
        if num > 2**31 - 1:
            return 2**31 - 1
        elif num < -2**31:
            return -2**31
        return num
        
            
"""
better solution
"""
class Solution:
    def myAtoi(self, s: 'str') -> 'int':
        if not str:
            return 0
        
        s = s.strip()
        if len(s) == 0:
            return 0
        
        sign = 1
        idx = 0
        if s[0] == "-":
            sign = -1
            idx = 1
        elif s[0] == "+":
            sign = sign
            idx = 1
        
        res = ""
        for n in s[idx:]:
            if n.isdigit():
                res += n
            else:
                break
                
        num = sign*int(res) if len(res) != 0 else 0
        if num > 2**31 - 1:
            return 2**31 - 1
        elif num < -2**31:
            return -2**31
        return num
        

File: 8. String to Integer (atoi)/8. String to Integer (atoi).js
/**
 * @param {string} str
 * @return {number}
 */
var myAtoi = function(str) {
    var i = 0, sign = 1, res = 0;
    
    while(str[i] === " ") {
        i += 1;
    }
    
    if (str[i] === "+") {
        i += 1;
        sign = sign;
    }
    else if (str[i] === "-") {
        i += 1;
        sign = -1;
    }
    
    for (; i < str.length; i++) {
        const code = str.charCodeAt(i) - 48; //"0" = 48
        if (code < 0 || code > 9) {
            break;
        }
        res = res*10 + code;
    }

    res = sign*res;
    return Math.max(-(2**31)-1, Math.min(2**31, res));
};
